purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/OfflineEntitlementsIntegrationTest.kt

```diff
@@ -0,0 +1,394 @@
+                        fail("Expected success but got error: $error")
+                    },
+                    onSuccess = { _, customerInfo ->
+                        assertCustomerInfoHasExpectedPurchaseData(customerInfo)
+                        assertAcknowledgePurchaseDidNotHappen()
+
+                        Purchases.sharedInstance.forceServerErrors = false
+
+                        Purchases.sharedInstance.getCustomerInfoWith(
+                            CacheFetchPolicy.FETCH_CURRENT,
+                            onError = {
+                                latch.countDown()
+                                fail("Expected success but got error: $it")
+                            },
+                            onSuccess = {
+                                // This is a known limitation. Ideally we would sync unsynced purchases
+                                // as soon as possible to avoid getting outdated info from the backend.
+                                assertCustomerInfoDoesNotHavePurchaseData(it)
+                                assertAcknowledgePurchaseDidNotHappen()
+
+                                latch.countDown()
+                            }
+                        )
+                    }
+                )
+            }
+        }
+    }
+
+    // region helpers
+
+    private fun setupTestWaitingForInitialRequests(
+        initialActivePurchases: Map<String, StoreTransaction> = emptyMap(),
+        postSetupTestCallback: (MainActivity) -> Unit = {}
+    ) {
+        setupTest(
+            initialActivePurchases = initialActivePurchases
+        ) {
+            waitForInitialRequestsToEnd {
+                postSetupTestCallback(it)
+            }
+        }
+    }
+
+    private fun waitForInitialRequestsToEnd(completion: () -> Unit) {
+        Purchases.sharedInstance.offlineEntitlementsManager.updateProductEntitlementMappingCacheIfStale {
+            Purchases.sharedInstance.getCustomerInfoWith(
+                onError = { fail("Expected to succeed") },
+                onSuccess = {
+                    completion()
+                }
+            )
+        }
+    }
+
+    private fun mockPurchaseResult(activePurchases: Map<String, StoreTransaction> = initialActivePurchases) {
+        every {
+            mockBillingAbstract.makePurchaseAsync(any(), any(), any(), any(), any(), any())
+        } answers {
+            mockActivePurchases(activePurchases)
+            latestPurchasesUpdatedListener!!.onPurchasesUpdated(activePurchases.values.toList())
+        }
+    }
+
+    private fun assertCustomerInfoDoesNotHavePurchaseData(customerInfo: CustomerInfo) {
+        assertThat(customerInfo.entitlements.active).isEmpty()
+        assertThat(customerInfo.activeSubscriptions).isEmpty()
+    }
+
+    private fun assertCustomerInfoHasExpectedPurchaseData(customerInfo: CustomerInfo) {
+        assertThat(customerInfo.entitlements.active.keys).containsExactlyInAnyOrderElementsOf(
```

Comment:
I wanted to note that this only works for these tests that are using offline entitlements. The token we currently are using for integration tests is for a expired purchase, that means that the backend won't return it as an active purchase. For these tests I'm mocking the purchase as returned from the Google method "queryPurchases" which returns active purchases, that's why it ends up as an active purchase there.

That's also the reason why I'm not verifying that the purchases are active after recovering from offline entitlements mode and only verifying that the purchase is acknowledged and sent to the backend.

This is not really great. I think we can improve this a lot by having an active purchase forever in a company credit card. Will study this option and try to improve this a bit.